SearchActivity.py

```diff
@@ -187,142 +180,64 @@ def _write_scores_to_clipboard(self, button=None):
         self._game.set_sharing(True)
 
     def _list_tubes_reply_cb(self, tubes):
         """ Reply to a list request. """
         for tube_info in tubes:
             self._new_tube_cb(*tube_info)
 
-    def _list_tubes_error_cb(self, e):
-        """ Log errors. """
-        _logger.debug('Error: ListTubes() failed: %s' % (e))
-
-    def _new_tube_cb(self, id, initiator, type, service, params, state):
-        """ Create a new tube. """
-        _logger.debug('New tube: ID=%d initator=%d type=%d service=%s \
-params=%r state=%d' % (id, initiator, type, service, params, state))
-
-        if (type == TelepathyGLib.TubeType.DBUS and service == SERVICE):
-            if state == TelepathyGLib.TubeState.LOCAL_PENDING:
-                self.tubes_chan[
-                    TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].AcceptDBusTube(id)
-
-            tube_conn = TubeConnection(
-                self.conn, self.tubes_chan[
-                    TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES], id,
-                group_iface=self.text_chan[
-                    TelepathyGLib.IFACE_CHANNEL_INTERFACE_GROUP])
-
-            self.chattube = ChatTube(tube_conn, self.initiating,
-                                     self.event_received_cb)
-
     def _setup_dispatch_table(self):
         ''' Associate tokens with commands. '''
         self._processing_methods = {
             'n': [self._receive_new_game, 'get a new game grid'],
             'p': [self._receive_dot_click, 'get a dot click'],
         }
 
-    def event_received_cb(self, event_message):
+    def _message_cb(self, collab, buddy, msg):
         ''' Data from a tube has arrived. '''
-        if len(event_message) == 0:
-            return
-        try:
-            command, payload = event_message.split('|', 2)
-        except ValueError:
-            _logger.debug('Could not split event message %s' % (event_message))
-            return
+        command = msg.get('command')
+        payload = msg.get('payload')
         self._processing_methods[command][0](payload)
 
     def send_new_game(self):
         ''' Send a new grid to all players '''
-        self.send_event('n|%s' % (json_dump(self._game.save_game())))
+        self.send_event('n', self._game.save_game())
 
     def _receive_new_game(self, payload):
         ''' Sharer can start a new game. '''
-        dot_list = json_load(payload)
+        dot_list = payload
         self._game.restore_game(dot_list)
 
     def send_dot_click(self, dot, color):
         ''' Send a dot click to all the players '''
-        self.send_event('p|%s' % (json_dump([dot, color])))
+        self.send_event('p', dot)
 
     def _receive_dot_click(self, payload):
         ''' When a dot is clicked, everyone should change its color. '''
-        (dot, color) = json_load(payload)
+        (dot, color) = payload
         self._game.remote_button_press(dot, color)
 
-    def send_event(self, entry):
+    def send_event(self, command, payload):
```

Comment:
You can also get rid of this method and simply have `collab.post` calls where you need them.

Reply:
Thanks